### ISSUE TITLE
Hide bridge driver implementation details from the host netns

### ISSUE THREAD

**Issue by corhere** [moby/moby#48222]
### Description

The bridge driver currently creates the Linux bridge devices in the host network namespace, which makes container-to-container traffic subject to the user's netfilter ruleset. Isolating container-to-container communication requires enabling the `net.bridge.bridge-nf-call-iptables` sysctl in the bridge's netns, [which can be problematic for users of virtual machines.](https://wiki.libvirt.org/Net.bridge.bridge-nf-call_and_sysctl.conf.html)

What we could do instead is place the bridge for the container network into an isolated netns so that all the container isolation netfilter rules and `net.bridge.bridge-nf-call-iptables` sysctl setting are isolated from the host. The bridge is then connected to the host network namespace over a "gateway" veth link.

- Internal network? Simply don't connect the gateway link to the host network. No need for DROP netfilter rules!
- Inter-bridge isolation? Just put them into different netns'es! The only route between two bridges is through the gateway veth, routed (not forwarded!) through the host netns, into the other bridge's gateway veth.
- Intuitive for users: they can apply any filtering, shaping, mangling, forwarding rules they want to traffic transiting the bridge's gateway veth link, just like any physical Ethernet link
  - No need to worry about breaking connectivity between containers connected to the same bridge

**Comment by zackherbert** [comment]
Related to #47828, it has been requested already for a long time.

**Comment by corhere** [comment]
> Related to #47828, it has been requested already for a long time.


Hmm, yes, you might be onto something! While this idea is neither necessary nor sufficient to afford providing containers a route to the internet via a netns of the user's choice, it would make a big difference in usability. The bridge and thus inter-container connectivity would behave consistently irrespective of the setup of the "gateway" netns, making it practical to give users the option to pick the gateway netns for a bridge network without it being a footgun.

**Comment by tianon** [comment]
Wouldn't this affect the (long existing) ability to connect to container IPs directly from the host, or would you envision a route through the gateway for the ranges?  How would that work for internal networks?

```console
$ docker network create --internal test
2d86c87ef6bc20ecfab4baf1c88c9f40a4c26bf09d9c2ae01d59bdf9cb0b70f1
$ docker run -d --name foo --network test debian sleep 100000
86b2f18ecce0b5ab257ee62a396f451934ad8a63169146e7c968d476b5faac59
$ docker network inspect -f '{{ json .Containers }}' test | jq '.[].IPv4Address'
"172.17.0.2/16"
$ ping -c1 172.17.0.2
PING 172.17.0.2 (172.17.0.2) 56(84) bytes of data.
64 bytes from 172.17.0.2: icmp_seq=1 ttl=64 time=0.154 ms

--- 172.17.0.2 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.154/0.154/0.154/0.000 ms
```

**Comment by corhere** [comment]
> Wouldn't this affect the (long existing) ability to connect to container IPs directly from the host

I don't think so. The gateway IP address would be assigned to the veth link in the host netns, so I am pretty sure that host<->container connectivity would continue to behave the same as today.

> How would that work for internal networks?

Oh right; I forgot that we use a subtly different meaning of "internal network" than everyone else. It would not work without a gateway veth link in the host netns. So we would still have to link the bridge to the host netns and use netfilter to isolate it from other networks. There goes one advantage. A bridge network without a gateway could be a new "really really isolated, I mean it" feature, though.

I may have oversold the advantages of such a change somewhat. I still suspect that it would be a net benefit, but further investigation and prototyping would be required to confirm that it's worth doing and that it won't break any existing use cases we'd want to support.
